### DISCUSSION TITLE
Show a creators posts & comments in their profile

### DISCUSSION THREAD

**Discussion by mubashariqbal** [deso-protocol/dips#93] [Medium CIPs]
## One-Line Description

Enable posts and comments to be returned by get-posts-for-public-key endpoint

## Explanation and Motivation

Currently only posts authored by a creator are displayed in their profile, any comments are only accessible thru the post they reply too.

We need an easy way to view a creators posts and comments to see all their postings.

## Specification

This will require updates to the core, backend and frontend to fully implement.  

I would like this CIP to focus on the core & backend changes to enable this functionality on the frontend.

A separate CIP should be created for the FE changes, which can be implemented/displayed in a number of different ways.

### Data Storage Change

None

## Backwards Compatibility

This will be an addition to existing functionality and as such there should be no compatibility issues.

**Comment by tijno** [comment]
In principle the idea seems sound and ive often wondered the same.

But there are 2-3 times the amount of comments vs main posts.

So hopefully someone smarter then me can assess impact on performance of adding comments into the response. 

And how performance of this can be optimised.

**Reply by mubashariqbal** [reply]
For the implementation I was thinking this would be a new parameter where comments **could** be enabled.

The existing functionality would stay the same, and so the impact on performance would be minimal.

For the front end a new tab could be added that shows "Posts & Comments", similar to how Twitter shows "Tweets & Replies".  This tab would access the same endpoint with the new parameter enabled to get the extra posts.

**Comment by mubashariqbal** [comment]
@maebeam @lazynina Would appreciate a review: approval or not, would be nice to get closure here 😀

**Reply by darian-parrish** [reply]
Given that this is a big part of the engagement on Twitter it seems wise to strongly consider this functionality.

**Comment by ItsAditya-xyz** [comment]
Great idea!
I believe we need another tab named "Replies" just like we have on twitter.  This will not mingle up the original posts with the replies. 

Also, I would want to see the post to which a user (whom I follow) has made an engagement in my following feed. 
This will be helpful in many ways:
- Discovery of new users
- more engagement

For example see this screenshot from twitter:
![Screenshot_20210909-091904_Chrome](https://user-images.githubusercontent.com/55331140/132619614-230bd7f2-e88a-4f26-b85d-5e26af9ebb6b.jpg)
I don't follow this user but since because Naval (whom I follow) has liked the post, the post is in my feed and hence Naval's engagement to that post has provided it more visibility. If the same thing had happened on Bitclout, no one would ever know about it. 

Also, this doesn't mean to show all the engaged posts. There needs to be a proper algorithm which will decide a set of users with whom I have been actively engaging. 
But, that's for another CIP....

**Comment by darian-parrish** [comment]
Not only do we need comments/replies on the user profiles but also in the following feed. This is significantly more challenging though due to the volume of comments that come through. 

Might have to get clever and splice them in based on diamonds/likes similar to the hot feed.

**Reply by matt-mcmarsh** [reply]
a reply to my poll
on deso https://desocialworld.com/posts/6d1968519471dc318b27e22a2721fa44c734fcf37489eab5d5028ea4c7b2bcf3

"A tab displaying all ‘re-activity’: all engagement activity you don’t see showing now on profile, basically.

Is my feedback/vote in?"

**Comment by matt-mcmarsh** [comment]
another from deso https://desocialworld.com/posts/cbcf9a80bc2b7e872ae20ff9b8fbbac4e6964daa95a17dcfb5651f8a2b322716

"[Yes but also would love a tab of everything except my reposts! (Could include quoted posts)]()"

**Comment by kitty4D** [comment]
i'd like to hear if this is or isn't possible from someone on core. i mean, it's ~ possible ~ of course, but i can understand that it may be too bone-crunching on .. i guess the nodes. what would make it much simpler? having some kind of list of the parent posthashhex's somewhere. 

> _sorry, i'm just like thinking of things maybe to help support making this feature possible. could there be some way to "build an index" that was stored on-chain that wouldn't interfere with anything, like posts that would contain all those parent posthashhex's containing comments so now you'd be looking at these new always-hidden posts that have some type of identifier in the extra data or something, find all of them that exist for the user being requested, each of these hidden posts has some number of hashes in the body.  maybe this doesn't make sense but just tossing it out here._  

but yeah absolutely being able to get a separate comments feed by user and/or one that mixes them amongst posts would be great to directly come from the api!

**Comment by jsonpreet** [comment]
Great Idea. There are more API routes we need to make dApps on DeSo. Let's see core team approves this DIP or not.

**Comment by diamondhands0** [comment]
I think it's about time we move this forward.

Let's do a $2,500 bounty for the best implementation that achieves the following:
* Backend endpoint to return all posts and replies
* Integrated into the deso-protocol/frontend UI: https://github.com/deso-protocol/frontend
* Integrated into the diamondapp/frontend UI: https://github.com/diamond-app/frontend

Any takers? Let us know!

**Comment by DeSoEngineer** [comment]
Just added two PRs to core and backend to add a "IncludeComments" boolean flag to the "get-posts-for-public-key" endpoint.

**Comment by donhardman** [comment]
I have implemented the requirements, and it should work 👀

https://github.com/deso-protocol/core/pull/418
https://github.com/deso-protocol/backend/pull/410
https://github.com/diamond-app/frontend/pull/248
https://github.com/deso-protocol/frontend/pull/614

If it is what you are looking for and I'm a "winner," I can improve the solution. Cheers.

**Comment by diamondhands0** [comment]
Hey guys, great work both of you!

After reviewing the code, @donhardman's implementation is the best so he wins the $2,500, but I'm going to send $500 in DESO to you as well @DeSoEngineer for the effort :)

@donhardman, we need to review the code a bit more before you refine the approach so just hang tight for now.

**Reply by darian-parrish** [reply]
Any progress on this?  I think this will be a quick way to increase engagement.

**Reply by StarGeezerPhil** [reply]
Same question as @darian-parrish - did this get merged? I thought it had, but only finally getting round to testing it and I'm not seeing any meaningful return from the API?

**Comment by matt-mcmarsh** [comment]
any idea when this may get actioned and included in nodes ?
